### ISSUE TITLE
User-created Custom Type Definition

### ISSUE THREAD

**Issue by Moderocky** [SkriptLang/Skript#4071]
TL;DR: A very simple way for users to define custom types and their simple property expressions.

### Introduction

A custom type would be effectively a Java class with property expressions attributed to its fields.
While this wouldn't allow for the same freedom as skript-mirror's custom expressions (or real syntax, for that matter!) it would give users a basic but perhaps more appropriate way of managing relevant data.

Types would have three essential components:
1. A type name (used for things like function argument types, or inside properties of other types)
2. Properties (equivalent to Java dynamic fields, these have a name and a type)
3. Core functions (such as how to stringify the type in quoted `"%thing%"` text, whether it can be saved, created, etc.)

Types would have to be parsed before actual Skript parsing occurs (similar to how function names/erasures are) so that they can be back-referenced and used in script files.

The basic layout is detailed below, though I am not married to the section names so better suggestions are welcome. :)
```c
type %string%: # the name
	properties:
		%string%: %typename% # the string key would be used in "the X of Y / Y's X" property name
	create:
		# trigger section allowing properties to be set
		# if present will create `[a] new %typename%` expression
	from string:
		# if present allows for parsing in `%string% parsed as %typename%`
	to string:
		# trigger section returns string
```

### Examples

In case this layout is unclear, an example of a custom type is below.
```c
type "elephant":
	properties:
		name: string
		age: number
		notes: strings # plural type so would be list
		weight: number
		parent: elephant # can allow reference to custom types here
	create:
		set name of this to "Jeremy" # assume 'this' to be self-variable
		set the age of this to 0
		add 1 to age of this # number type so we can use simple changers
		add "New elephant!" to notes of this # list type so we can use add/remove changers
		set weight of this to (random number between 100 and 200)
	to string:
		return "%name of this% the Elephant"
	from string: # the 'create' section would be run first to set 'default' values
		set {_s::*} to event-string split at " the Elephant"
		set name of this to {_s::1}
```

The properties of the type would have auto-generated property expressions to allow the user to manipulate them easily in their code. An example is below.

```c
on join:
	set {_a} to a new elephant
	set {_a}'s name to "Bob"
	broadcast "%{_a}%" # invokes 'to string' -> 'Bob the Elephant'
	set {_b} to ("%{_a}%" parsed as elephant) # invokes 'create' then 'from string'
	set {_a}'s age to 100
```

Naturally, custom types would be entirely non-functional and would exist mainly as a way to give users a more ordered and structured way to deal with commonly-associated data rather than just creating list variables for them.

This would also give users who create Skript "libraries" an easier way to handle their data and make it much neater for organising.

### Parsing and Serialisation

As long as all of the property types of a custom type are serialisable, then the custom type will be too.
Types that are not, or are partly serialisable (perhaps one property stores a value that can't be saved to `variables.csv` can warn the user if they store it in a global variable as Skript does already.

### Class Creation

Types will generate a class at runtime, which will be used for registering the Skript type.
The type properties will be bound to the class's dynamic fields, and we can extract their type from the property classinfo.

Generating the bytecode for the type class shouldn't be problematic, since all it will need is simple field declarations and the rest can be handled in the syntax generation (which I have already written tools for)

The only decision would be whether to keep the type classes at a defined path (something like `ch/njol/skript/generated/CustomTypes$typename`), so they would need disposing of upon reload but would be much easier to backreference since the classpath *is already known* before the type is even created, or whether to assign them a randomly-generated name (like `ch/njol/skript/generated/CustomTypes$Type1283289`), which would minimise any potential conflicts but make handling backreferences and self-references much more complicated.
I would personally recommend the former since it's less effort, but either would work.

Since all names and types would be known at generation time, there would be no need for reflection other than to allocate type instances (`new %thing%`) - all fields can be properly invoked in the generated code so runtime overhead is minimised.

Properties would likely need their own syntax classes registering, but I've dealt with this before in other projects so can definitely provide all the utilities for this if needed - it's not too difficult.

When handling name conflicts (such as two scripts registering a type with the same name) an error would have to be thrown, but since this would happen with functions or addons registering identical things I don't see this as being a major issue.
If anything, it's less of an issue since a custom type would be open to and editable by the user.

The functional `create:` member can either be run from the constructor of the type class (I'm not sure how to hook a skript function from Java, but I assume it's possible to create something like that here?) or could instead be triggered by Skript itself whenever the `[a] new %customtype%` syntax is used. In either case, type classes with no `create:` section would need no constructor in the class since there would be no assignments made.

The functional `to string:` member can be referenced from the object's `toString` method if the code for it is handled in Skript. I think the `from string:` member would need to be deferred to the type, but I'm not sure how Skript handles that internally.

The only potential issue I can see here are script reloads, and perhaps the user could be warned that types cannot be altered between full server reloads (since they register types, classes and syntax that need to be properly ceded and garbage-collected to avoid memory leaks)

### Semantics

I am not keen on adding type-specific effects since Skript is not an object-oriented language like Java, and we already have functions for that.

While one might argue that custom types would add nothing that cannot already be done through list variables, I think this is a very weak criticism. By the same logic you could do away with Java's objects and fields and work with their data in maps instead. This is not supposed to add some major new functionality but instead clean up Skript and give users a neater and more ordered way to manage data relations.

This is not at all a replacement for list variables (and in 90% of cases a list variable will be more appropriate than a custom type) but in the other ten percent of cases it will give users a safe and easy way to keep their work organised and tidy, especially if it is something they are offering as a "library" or utility.

Think of custom types as like Java records, structs or typeless JavaScript objects. They would exist much more as an organisational tool than something to actually add new functionality (though it may well be that they do provide unintended uses.)

**Comment by AyhamAl-Ali** [comment]
Just from the title itself I LIKE IT 🔥 Well described.
